### ISSUE TITLE
spring-boot-loader-tools launch.script on some systems writes output logs to $pid_file

### ISSUE THREAD

**Issue by bedge** [spring-projects/spring-boot#3742]
I thought ops was crazy too at first, but after looking at it, I'm at a bit of a loss. I think it may be related to argument quoting on older su commands, but the end result is that it runs this command:
```
su -c "$command &> \"$log_file\" & echo \$!" $run_user > "$pid_file"
```
On older linux VMs, in this case Ubuntu 12.04, it ends up writing the logs to the $PID file:
```
-rw-r--r-- 1 nim_cas root 68231 Aug 12 20:48 /var/run/app/app.pid
```
I've compared systems back to back with set -x on the init.d/ script and the commands they run are identical. I believe it's related to the older login package that provides the su command.

Here's the /proc/<PID>/fd of a "normal" system:
```
lr-x------ 1 nim_cmas nim_cmas 64 Aug 12 21:09 0 -> /dev/null
l-wx------ 1 nim_cmas nim_cmas 64 Aug 12 21:09 1 -> /var/log/app.log
l-wx------ 1 nim_cmas nim_cmas 64 Aug 12 21:09 2 -> /var/log/app.log
```
and here's a problem system,  note that fd 1 = pid file.
```
lr-x------ 1 nim_cas nim_cas 64 Aug 12 21:18 0 -> /dev/null
l-wx------ 1 nim_cas nim_cas 64 Aug 12 21:18 1 -> /var/run/app/app.pid
lrwx------ 1 nim_cas nim_cas 64 Aug 12 21:17 2 -> /dev/pts/4
```

**Comment by bedge** [comment]
I believe this may work:
```
    su -c "$command &> \"$log_file\"" & echo $!  $run_user > "$pid_file"
```
in place of the orignial
```
    su -c "$command &> \"$log_file\" & echo \$!" $run_user > "$pid_file"
```

**Comment by bedge** [comment]
Just in case anyone else runs into this, it appears to be bug in the older version of the su utility that ships with Ubuntu versions prior to 14.04.

**Comment by shakuzen** [comment]
This sounds like something that might be worth documenting.

**Comment by philwebb** [comment]
@bedge @shakuzen Do you know of any workarounds? Can we change that command in the script somehow?

**Comment by bedge** [comment]
Haven't found anything, but I haven't had time to spend on it either.
I briefly looked at converting the init.d script to using start-stop-daemon, which does not suffer from tis problem, but that's more than 5 minute job.

**Comment by bedge** [comment]
This just started happening on 14.04 systems as well. Not sure what changed recently but I'm pretty sure it was a spring update as it also broke the yaml config reading, and I need to manually add snakeyaml to runtime dependencies now.
This happened sometime between Oct 2 - Oct 8.
Anyway, this problem is not restricted to pre-14.04 ubuntu systems now.

**Comment by bedge** [comment]
Looking at the changes, I believe 3865f3b is the culprit.

Would anyone consider a pull if a submitted a patch that replaced su with start-stop-daemon in the launch.script?

**Comment by bedge** [comment]
Submitted a pull request that resolves this: https://github.com/spring-projects/spring-boot/pull/4137

**Comment by bedge** [comment]
Please let me know if this would be preferable:

Currently this change will use start-stop-daemon if present. Another option would be to set a flag in ${jarfolder}/${configfile} to configure which start mechanism to use.
eg:
in ${jarfolder}/${configfile}:

export USE_START_STOP_DAEMON=1

Then, rather than just checking the existence of start-stop-daemon, do:

```
if [ "$USE_START_STOP_DAEMON" ] ; then
    use it
else
    use current su
fi
```

as opposed to the current pull req:
```
if  which start-stop-daemon >/dev/null ; then 
    use it
else
    use current su
fi
```

**Comment by philwebb** [comment]
Fixed in df07151f10b6832ae235160ce070258cf923e36a
